### ISSUE TITLE
Improve typing of FilterQuery

### ISSUE THREAD

**Issue by vimmerru** [mikro-orm/mikro-orm#124]
**Describe the bug**
You can put to ```em::find``` anything as $where condition that completely breaks any compile time checking of queries. It would be nice to keep typing at least for simple cases as for example TypeORM provides.

**Comment by B4nan** [comment]
Well the problem here is that `where` also allows to pass smart query conditions, e.g. for entity with `age` property the where can be also `{ 'age:gte': 18 }` .

Not sure how to approach this, can you think of a way to accept smart query conditions as well as preserve the original entity type?

**Comment by vimmerru** [comment]
> Not sure how to approach this, can you think of a way to accept smart query conditions as well as preserve the original entity type?

I suggest the following define types:

```Query<T>``` and ```SmartQuery``` and allow the following type as query condition (isn't correct typescrypt, but illustrates the idea):

```
query: DeepPartial<T> | {$where: Query<T>} | {$smartWhere: SmartQuery}
```

**Comment by vimmerru** [comment]
the main idea that by default is just set of entity fields, but you can also use field $where that allow to write typed queries or field $smartWhere that allow to write untyped smart queries.

**Comment by B4nan** [comment]
Yeah that would be a possible way, although breaking change and with worse API (but type safe where possible). 

What would be `Query<T>` and `SmartQuery` types defined like? I guess `SmartQuery` would be still just `Record<string, any>`? And `Query<T>` would have those operators added to `T` properties?

I will try to look into possibilities of using generated types as that would allow to keep current API, but strongly typed (generated types would have all possible query operators included for each property). It would be still breaking change most probably, and I am still not sure if its even possible :]

**Comment by vimmerru** [comment]
> Yeah that would be a possible way, although breaking change and with worse API (but type safe where possible).

From my point of view safe em queries are much more important and welcome than beautiful API ), but it just IMHO.

> I am still not sure if its even possible :]

There was proposal about strongly typed QueryBuilder for TypeORM based on typescripts types only. Also there is https://github.com/Ff00ff/mammoth

**Comment by vimmerru** [comment]
I want to spend some time for PoC, but now i am really overloaded.

**Comment by B4nan** [comment]
> > I am still not sure if its even possible :]
> 
> There was proposal about strongly typed QueryBuilder for TypeORM based on typescripts types only. Also there is https://github.com/Ff00ff/mammoth

What I meant is to have strict types including those smart query operators... That would be possible only with generated types I guess. 

Or maybe we could let users whitelist additionally allowed properties, something like:

```typescript
type FilterQuery<T, K> = Partial<T> & { [TK in K]: any };
const where: FilterQuery<Author, 'age:gte'> = { age: '123', 'age:gte': 18 };
```

And having it by default (without the `K` parameter) require only valid entity props. In the end one can always do `const where: FilterQuery<Author> = { name: { $gte: 18 };` which would be even shorter that the previous version with `K` parameter.

I think I could just remove that `age:gte` syntax and allow only the object one (which should be easily type-able). I agree that type-safety is more important that syntax sugar, especially this one that is easily rewritten to type-checkable form. 

After few hours of prototyping, here is a proof of concept:

```typescript
    type Author = { name: string; age: number; born: Date };
    type GroupOperator = keyof typeof QueryBuilderHelper.GROUP_OPERATORS;
    type Operator = keyof typeof QueryBuilderHelper.OPERATORS | GroupOperator;
    type OperatorMap<T> = { [OP in Operator]?: T | T[] };
    type GroupOperatorMap<T> = { [OP in GroupOperator]?: T | T[] };
    type Query<T> = { [K in keyof T]? : OperatorMap<T[K]> | T[K] };
    type FilterQuery<T> = GroupOperatorMap<Query<T>> & Query<T>;
    const where: FilterQuery<Author> = { age: { $gte: 18 }, born: { $lt: new Date() }, $and: [{ name: { $ne: 'John' } }, { name: { $in: ['Ben', 'Paul'] } }] };
```

There will be definitely still some cases not covered, but a good start I would say :] I will change this into feature issue and include this in v3 as its gonna be breaking change.

http://www.typescriptlang.org/play/#code/KYDwDg9gTgLgBAYwDYEMDOa4EUCuwoCeAQjgJZIAm+AEsEmPnAN4BQccaMKMpCcUwFBQgA7JATgBxAEoB5AKoAFAPqzFAUWkBBACqzpAZTgBeZm3ZwAJChEUAXHADkNio4A059pegPH0d+YAvgDc5pzcvPyCwmISapq6+kamrBZWwACOvsYBaZakIr4FuRaWIgW+IhDwxR55AOYwvgB8JV6NwC05daVITU4APG1WfZ2D3Z5WImOOAIQT7CEsgSwsMAQMcFo4MAAW0CbMcCIoALZjnFAF9cFwKPVjIjinAEb4ty-QhXAAItzAcCW602kigEBwYFkDCg3AOpgA1sACBAAGZwYHAVHYPCEEjkKhQWj0fAAOhkChU8W0ekMoQxcCh+FhUEOiORaIxWNw+GIZEoNDo0JJVMShjgAB8pGCIYyYTBoHSNgDZcyALIoMADHTNQ5MOAAbTUcAKDOhzIAugB+Bw6CVwHT682AxUg6WQs3yqDqzXa3UGo0m0Hg91Mz1Wm12h1OoFK7E8rU6lIGgDSxpEcDZWJ0VrgDhVnu9Wv1yfNOslDpLzrWsYAYuQYPhuYQE4cgzKPdBC02CAmdQAyOPN7WhFgAelHcAAbigkKQKCwEKJOHAAO67fDAAAMDjroyg3YG2z20ETR3uYz1lg6DgAjAAOQFVxciZdrjc3nf1xs4ntH-ZQU89XPBxL2vOB70BNw4E+KBvkvPoHGmFdfn+AAKABKSCrBcBx9T1E5zhAqYZgAKQgXYREcR9Aig-CzgvKwKgNRwiGASioMcRQUBwJBHGjQFoxHcc02nWd52fV91wEAAmT89wPP8Tz9c8b2kgBmW8HyWCT4DfAQNLgXcG33H9Dx2f9ALuB4iKvBtfHvKjAifJddKk4AABY5OMhTzKUpMYLgqwwMcVS1Mc5yX1cjcAFYvO-eNFIA5TrKOWyZgcrCAt8Cg0BQdxsNsXC6MI1LplvdTqNo456Jspj9TUjiuJ4vjqMElggA

**Comment by vimmerru** [comment]
It would be nice to get some workaround for reference fields. It is common to use

```class User {
   organization: Organization
}

find(User, { organization: 123 }}
```

It can be solution based on DeepPartial<T> like 


```class User {
   organization: Organization
}

find(User, { organization: { id: 123} }}
```

**Comment by B4nan** [comment]
Agreed, this will need to be supported. I will try to come up with a way to preserve current way with directly passing the PK first.

**Comment by B4nan** [comment]
I think I finally cracked this one. Need to clean it up but this is working already:

```typescript
class Organization {
  id: number;
  name: string;
}
interface Organization extends IdentifiedEntity<Organization> { }

class Author {
  id: number;
  name: string;
  age: number;
  born: Date;
  organization: Organization;
}
interface Author extends IdentifiedEntity<Author> { }

// valid
const where1: FilterQuery<Author> = { age: { $gte: 18 } };
const where2: FilterQuery<Author> = { age: { $gte: 18 }, born: { $lt: new Date() }, $and: [{ name: { $ne: 'John' } }, { name: { $in: ['Ben', 'Paul'] } }] };
const where3: FilterQuery<Author> = { organization: new Organization() };
const where4: FilterQuery<Author> = { $and: [{ organization: new Organization() }, { name: 'Bob' }] };
const where5: FilterQuery<Author> = new Author();
const where6: FilterQuery<Author> = 1;
const where7: FilterQuery<Author> = { organization: 1 };

// invalid
const invalid1: FilterQuery<Author> = { age123: 18 }; // prop not exist
const invalid2: FilterQuery<Author> = { age: { $gte: '18' } }; // wrong type, age is number
const invalid3: FilterQuery<Author> = { born: { $gte: '123' } }; // wrong type, born is Date
const invalid4: FilterQuery<Author> = { age: { $gte: 18 }, $and: [{ name: { $ne: '123' } }, { name: { $in: [3, 'Paul'] } }] }; // name should be string, checks also values in the array
const invalid5: FilterQuery<Author> = { age: { $gte: '18' }, $and: [{ name: { $ne: '123' } }, { name: { $in: ['3', 'Paul'] } }] }; // age should be number
const invalid6: FilterQuery<Author> = '1'; // wrong type of PK, id is number
const invalid7: FilterQuery<Author> = new Organization(); // wrong class, should be Author
const invalid8: FilterQuery<Author> = { organization: new Author() }; // wrong property class, should be Organization
const invalid9: FilterQuery<Author> = { organization: '1' }; // wrong property PK type, Organization.id is number
```

**Comment by vimmerru** [comment]
What is the definition of FilterQuery?

**Comment by B4nan** [comment]
```typescript
export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
    ? DeepPartial<U>[]
    : T[P] extends ReadonlyArray<infer U>
      ? ReadonlyArray<DeepPartial<U>>
      : DeepPartial<T[P]>
};

export type Primary<T> = T extends { __primaryKey: infer PK } ? PK : never;

type GroupOperator = keyof typeof QueryBuilderHelper.GROUP_OPERATORS;
type Operator = keyof typeof QueryBuilderHelper.OPERATORS | GroupOperator;
type OperatorMap<T> = { [OP in Operator]?: DeepPartial<T> | DeepPartial<T>[] };
type GroupOperatorMap<T> = { [OP in GroupOperator]?: DeepPartial<T> | DeepPartial<T>[] };
type StringProp<T> = T extends string ? string | RegExp : never;
type EntityOrPrimary<T> = Partial<T> | DeepPartial<T> | Primary<T>;
type CollectionItem<T> = T extends Collection<infer K> ? EntityOrPrimary<K> : never;
type FilterValue<T> = OperatorMap<T> | DeepPartial<T> | StringProp<T> | EntityOrPrimary<T> | EntityOrPrimary<T>[] | CollectionItem<T> | CollectionItem<T>[];
type Query<T> = { [K in keyof T]?: (T[K] extends IEntity ? Query<T[K]> : FilterValue<T[K]>) | null } | FilterValue<T>;

export type FilterQuery<T> = GroupOperatorMap<Query<T>> & Query<T>;
```

But still there are some edge cases not covered.

**Comment by vimmerru** [comment]
What do you think if we force explicit usage of Ref type for entities definition like

```
class Bar {
}

class Foo {
  bar1: Ref<Bar>;
  bar2: Ref<Bar>;
}
```

It can be possible to derive type with resolved references

```
type DeRef<T> = T extends Ref<infer U> ? U : T;

type DeRefObject<T> = {[k in keyof T]: DeRef<T[k]>}

type DeRefPartial<T, K extends keyof T> = {
  [k in keyof T]: k extends K ? DeRef<T[k]> : T[k]
}
```

This type can be used as basis of FilterQuery. Population can be done by de-referencing some path. Unfortunately there is no generic solution for this, but if we limit how deep population is the solution exists. 

It will allow typed queries as well as typed population.

**Comment by B4nan** [comment]
How would then look the actual query/find call and the population?

**Comment by vimmerru** [comment]
> How would then look the actual query/find call and the population?

Let's consider reference definition and de-referenced type like this

```
class Ref<T> {
  async unwrap<T>(): Promise<T> {
    return {} as any;
  }
}

type DeRef<T> = T extends Ref<infer U> ? U : T;
```

We can defined type Populated<T> like this (for now just one level of population):

```
type Populated<T, K extends keyof T> = {
  [k in keyof T]: k extends K ? DeRef<T[k]> : T[k]
}
```

If we have model

```
class Credentials {
}

class Address {
}

class User {
  credentials: Ref<Credentials>;
  address: Ref<Address>;
}
```

Type ```Populated<User, 'credentials'>``` will be resolved as

```
class User {
  credentials: Credentials;
 address: Ref<Address>;
}
```

Type ```Populated<User, 'credentials' | 'address'>``` will be resolved as

```
class User {
  credentials: Credentials;
  address: Address;
}
```


```em.find(User, $were)``` will return ```List<User>``` but we can have an API like:

```
await em.find(User, $were)
   .populate('credentials')
   .populate('address')
   .from($from)
   .limit($limit)
   .exec()
```

this request will return ```List<Populated<User, 'credentials' | 'address'>>```

**Comment by vimmerru** [comment]
The real problem is definition of DeepPopulate type based on nested path, but it can be solved by limiting nesting and definition of types Populate2, Populate3 and etc... ```populate``` function can have multiple signatures based on different about of parameters.

**Comment by B4nan** [comment]
I think I got it working pretty well, but its so complex that the mysql EM test is failing in the CI :D But that can be fixed by simply splitting the test, its really huge and lot of FilterQuery assertions are made there...

**Comment by B4nan** [comment]
Closing this now, available in `3.0.0-alpha.23`. Please open new issues for bugs.
